### PR TITLE
Support deletion in function `translate` 

### PR DESCRIPTION
**Pull request by shuai-xu**
### Changelog category (leave one):
- Improvement

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Function `translate` now supports character deletion if the `from` argument contains more characters than the `to` argument. Example: `SELECT translate('clickhouse', 'clickhouse', 'CLICK')` now returns `CLICK`.

### CODE REVIEW

**Comment by robot-ch-test-poll3** [comment]
*This is an automated comment for commit 42e53f9132f411aef5f842359f97bdabd92a2445 with description of existing statuses. It's updated for the latest CI running*

[✅ Click here](https://s3.amazonaws.com/clickhouse-test-reports/71441/42e53f9132f411aef5f842359f97bdabd92a2445/ci_running.html) to open a full report in a separate page

<details><summary>Successful checks</summary>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>Style check</td><td>Runs a set of checks to keep the code style clean. If some of tests failed, see the related log from the report</td><td><a href="https://s3.amazonaws.com/clickhouse-test-reports/71441/42e53f9132f411aef5f842359f97bdabd92a2445/style_check.html">✅ success</a></td></tr>
<tbody>
</table>
</details>

**Review by yariks5s — COMMENTED** [review]
Hi! Actually, I cannot understand why we need this logic for the `translate` function.

As far as I understood, if we have `translate(string, 'abc', 'ab')`, it will be identical to `translate(string, 'ab', 'ab')`, right? If yes, why do we need to specify an extra letter If we can just not use this extra letter?

**Comment by shuai-xu** [comment]
> Hi! Actually, I cannot understand why we need this logic for the `translate` function.
> 
> As far as I understood, if we have `translate(string, 'abc', 'ab')`, it will be identical to `translate(string, 'ab', 'ab')`, right? If yes, why do we need to specify an extra letter If we can just not use this extra letter?

Thank you for reply. `translate(string, 'abc', 'ab')` is not identical to `translate(string, 'ab', 'ab')`. For `translate(string, 'abc', 'ab')` ， the letter c in string will be translated to empty. In our production, users may use translate(string, '\"', '') to remove the letters.

**Comment by yariks5s** [comment]
Okay, thank you, now it makes sense for me

File: src/Functions/translate.cpp [hunk 1]
```diff
@@ -59,29 +72,39 @@ struct TranslateImpl
         fillMapWithValues(map, map_from, map_to);
 
         res_data.resize(data.size());
-        res_offsets.assign(offsets);
+        res_offsets.resize(input_rows_count);
```
[the remark this answers is not in the document]
Comment (shuai-xu) [reply src/Functions/translate.cpp:75]: The result string's length may be not same with the from string, so the offset of dst need to be calculated later.

File: src/Functions/translate.cpp [hunk 2]
```diff
@@ -59,29 +72,39 @@ struct TranslateImpl
...
                     *dst = map[*src];
-                else
+                    ++dst;
+                    ++data_size;
+                }
+                else if (*src > ascii_upper_bound)
+                {
                     *dst = *src;
+                    ++dst;
+                    ++data_size;
```
[the remark this answers is not in the document]
Comment (shuai-xu) [reply src/Functions/translate.cpp:97]: If `*src <= ascii_upper_bound && map[*src] == ascii_upper_bound + 1`, which means the from str is valid, but the letter should be translated to empty, the dst should not add 1.

File: src/Functions/translate.cpp [hunk 3]
```diff
@@ -92,6 +115,9 @@ struct TranslateImpl
         const std::string & map_to,
         ColumnString::Chars & res_data)
     {
+        if (map_from.size() != map_to.size())
+            throw Exception(ErrorCodes::BAD_ARGUMENTS, "Second and third arguments must be the same length");
+
```
[the remark this answers is not in the document]
Comment (shuai-xu) [reply src/Functions/translate.cpp:120]: Yes we can support it. But for FixedString, if we allow it, the result's length may be not a fixed string, I am not sure whether users need this. So for FixedString, I suggest not support it now. What do you think?

File: src/Functions/translate.cpp [hunk 4]
```diff
@@ -32,17 +32,30 @@ struct TranslateImpl
...
-            map[map_from[i]] = map_to[i];
+                map[map_from[i]] = map_to[i];
+            }
+            else
+            {
+                if (!isASCII(map_from[i]))
+                    throw Exception(ErrorCodes::BAD_ARGUMENTS, "Second argument must be ASCII strings");
+
+                map[map_from[i]] = ascii_upper_bound + 1;
+            }
```
[the remark this answers is not in the document]
Comment (shuai-xu) [reply src/Functions/translate.cpp:52]: Sorry, I am not sure what do you mean?

**Comment by shuai-xu** [comment]
@yariks5s can you help to review this pr again?

**Comment by yariks5s** [comment]
Yes, I will, sorry for the delay

File: docs/en/sql-reference/functions/string-replace-functions.md [hunk 5]
```diff
@@ -253,7 +253,7 @@ SELECT format('{} {}', 'Hello', 'World')
 
 ## translate
 
-Replaces characters in the string `s` using a one-to-one character mapping defined by `from` and `to` strings. `from` and `to` must be constant ASCII strings of the same size. Non-ASCII characters in the original string are not modified.
+Replaces characters in the string `s` using a one-to-one character mapping defined by `from` and `to` strings. `from` and `to` must be constant ASCII strings. Non-ASCII characters in the original string are not modified.
```
Comment (yariks5s) [inline docs/en/sql-reference/functions/string-replace-functions.md:257]:
```suggestion
Replaces characters in the string `s` using a one-to-one character mapping defined by `from` and `to` strings. `from` and `to` must be constant ASCII strings. Non-ASCII characters in the original string are not modified. If the number of characters in `from` list, overlapping characters will be deleted from the input string.
```

File: src/Functions/translate.cpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (yariks5s) [reply src/Functions/translate.cpp:52]: If we have |`map_from`| > |`map_to`|, we don't need to check the arguments of `map_to` again.

```suggestion
            else
            {
                while (i < map_from.size())
                {
                    if (!isASCII(map_from[i]))
                        throw Exception(ErrorCodes::BAD_ARGUMENTS, "Second argument must be ASCII strings");

                    map[map_from[i]] = ascii_upper_bound + 1;
                    i++;
                }
                return;
            }
```

**Review by yariks5s — COMMENTED** [review]
Also, organizational note:

Should we support the case if |`to`| > |`from`|? 
If not, we don't need some part of logic implemented in this PR (checking overlapping values of `map_to`)

**Comment by yariks5s** [comment]
Tbh, I have a feeling that we can do https://github.com/ClickHouse/ClickHouse/pull/71441#discussion_r1852703207 even better.

**Comment by shuai-xu** [comment]
> Tbh, I have a feeling that we can do [#71441 (comment)](https://github.com/ClickHouse/ClickHouse/pull/71441#discussion_r1852703207) even better.

Thank you for review again, I have modify it according to your comments.

File: src/Functions/translate.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (yariks5s) [reply src/Functions/translate.cpp:102]: At least we can do it in the way it was before:

``` C++
            while (src < src_end)
            {
                if (*src <= ascii_upper_bound && map[*src] != ascii_upper_bound + 1)
                {
                    *dst++ = map[*src];
                    ++data_size;
                }
                else if (*src > ascii_upper_bound)
                {
                    *dst++ = *src;
                    ++data_size;
                }

                ++src;
            }
```

File: src/Functions/translate.cpp [hunk 6]
```diff
@@ -226,6 +268,12 @@ struct TranslateUTF8Impl
                         auto * it = map.find(*src_code_point);
                         if (it != map.end())
                         {
+                            if (it->getMapped() == max_uint32)
+                            {
+                                src += src_len;
+                                continue;
+                            }
+
```
Comment (yariks5s) [inline src/Functions/translate.cpp:276]: Here you can use only one `src += src_len;` statement for the whole if-statement:
```suggestion
                            src += src_len;
                            if (it->getMapped() == max_uint32)
                                continue;
```

and please remove the `src += src_len;` below in this if-statement

File: tests/queries/0_stateless/02353_translate.sql [hunk 7]
```diff
@@ -10,4 +10,10 @@ SELECT translate('abc', '', '');
...
 
 SELECT translate('abc', 'Ááéíóúôè', 'aaeiouoe'); -- { serverError BAD_ARGUMENTS }
-SELECT translateUTF8('abc', 'efg', ''); -- { serverError BAD_ARGUMENTS }
+SELECT translateUTF8('abc', 'efg', '');
+
+SELECT translateUTF8('中文内码', '中文', '');
+SELECT translate('aAbBcC', 'abc', '12');
+
+SELECT translate('aAbBcC', 'abc', '1235');
+SELECT translate('aAbBcC', '中文内码', '12'); -- { serverError BAD_ARGUMENTS }
```
Comment (yariks5s) [inline tests/queries/0_stateless/02353_translate.sql:19]: Could you also please add a test like:

SELECT translate('abc', 'abc', '');

**Review by yariks5s — APPROVED** [review]
Please leave your thoughts about it if you have some, these changes are not mandatory but will improve the readability of the code, and sorry for the long responses :(

**Comment by yariks5s** [comment]
+ Please update your pull request with master, because we have some tests fixed on master.

**Comment by shuai-xu** [comment]
@yariks5s Thank you very much , all comments are addressed.

File: src/Functions/translate.cpp [hunk 8]
```diff
@@ -32,17 +32,35 @@ struct TranslateImpl
...
+                }
+                return;
+            }
+        }
+
+        for (size_t i = map_from.size(); i < map_to.size(); ++i)
+        {
+            if (!isASCII(map_to[i]))
+                throw Exception(ErrorCodes::BAD_ARGUMENTS, "Third argument must be ASCII strings");
         }
```
[the remark this answers is not in the document]
Comment (shuai-xu) [reply src/Functions/translate.cpp]: done

File: tests/queries/0_stateless/02353_translate.sql (hunk 7 above)
Comment (yariks5s) [reply tests/queries/0_stateless/02353_translate.sql:20]: I think you forgot to add the test to commit

**Comment by yariks5s** [comment]
Overall except the issue about the test is good, thank you🙂

File: tests/queries/0_stateless/02353_translate.sql (hunk 7 above)
Comment (shuai-xu) [reply tests/queries/0_stateless/02353_translate.sql:21]: sorry，done now

**Comment by shuai-xu** [comment]
> Overall except the issue about the test is good, thank you🙂

@yariks5s Thank you, test added now。

File: src/Functions/translate.cpp [hunk 9]
```diff
@@ -32,17 +32,31 @@ struct TranslateImpl
         const std::string & map_from,
         const std::string & map_to)
     {
-        if (map_from.size() != map_to.size())
-            throw Exception(ErrorCodes::BAD_ARGUMENTS, "Second and third arguments must be the same length");
-
         iota(map.data(), map.size(), UInt8(0));
 
-        for (size_t i = 0; i < map_from.size(); ++i)
+        size_t min_size = std::min(map_from.size(), map_to.size());
```
Comment (rschu1ze) [inline src/Functions/translate.cpp:37]: The case that the third argument is longer than the second argument is weird, it should throw an exception: `select translate('abc', 'ab', 'abc')`.  @yariks5s ?

Comment (yariks5s) [reply src/Functions/translate.cpp:37]: https://github.com/ClickHouse/ClickHouse/pull/72893

**Pull request merged**